src/Psy/ExecutionClosure.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * This file is part of Psy Shell.
+ *
+ * (c) 2012-2017 Justin Hileman
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Psy;
+
+/**
+ * The Psy Shell's execution scope.
+ */
+class ExecutionClosure
+{
+    const NOOP_INPUT = 'return null;';
+
+    private $closure;
+
+    /**
+     * @param Shell &$__psysh__
+     */
+    public function __construct(Shell &$__psysh__)
+    {
+        $exec = function () use (&$__psysh__) {
+            try {
+                // Restore execution scope variables
+                extract($__psysh__->getScopeVariables(false));
+
+                // evaluate the current code buffer
+                ob_start([$__psysh__, 'writeStdout'], 1);
+
+                set_error_handler([$__psysh__, 'handleError']);
+                $_ = eval($__psysh__->onExecute($__psysh__->flushCode() ?: ExecutionClosure::NOOP_INPUT));
+                restore_error_handler();
```

Comment:
Probably want this in a finally?

Reply:
I did think about this when cleaning this up. I'm not sure we do. `handleError` should only handle errors during `eval` (or as close to it as possible). If we put it in a finally, it'll handle errors in the flush/clean, as well as errors in setting scope variables. We could narrow it down to just handling errors in eval and the `ob_end_flush`, which is probably good enough…

```diff
diff --git a/src/Psy/ExecutionClosure.php b/src/Psy/ExecutionClosure.php
index 7dfd7eb..5e57d12 100644
--- a/src/Psy/ExecutionClosure.php
+++ b/src/Psy/ExecutionClosure.php
@@ -32,30 +32,20 @@ class ExecutionClosure
 
                 // evaluate the current code buffer
                 ob_start([$__psysh__, 'writeStdout'], 1);
-
                 set_error_handler([$__psysh__, 'handleError']);
                 $_ = eval($__psysh__->onExecute($__psysh__->flushCode() ?: ExecutionClosure::NOOP_INPUT));
-                restore_error_handler();
 
+                // Has to be done before the finally or it'll get cleaned rather than flushed.
                 ob_end_flush();
-
-                $__psysh__->setScopeVariables(get_defined_vars());
-            } catch (\Throwable $_e) {
+            } finally {
                 restore_error_handler();
                 if (ob_get_level() > 0) {
                     ob_end_clean();
                 }
-
-                throw $_e;
-            } catch (\Exception $_e) {
-                restore_error_handler();
-                if (ob_get_level() > 0) {
-                    ob_end_clean();
-                }
-
-                throw $_e;
             }
 
+            $__psysh__->setScopeVariables(get_defined_vars());
+
             return $_;
         };
 ```

... but it's all a moot point because `finally` is PHP 5.5+ and this version of PsySH is still targeting >= 5.4 :(